### ISSUE TITLE
git-hack: always push the new branch to the remote

### ISSUE THREAD

**Issue by kevgo** [Originate/git-town#664]
* allows `git push` on the new branch right away
* tells other team members that this branch exists now
* prerequisite for #663

**Comment by kevgo** [comment]
blocked by #519

**Comment by allewun** [comment]
Perhaps a contrived problem, but hacking a new branch can cause CI (or other git hook services) to spin up and build a branch that has no new commits. In repos that have long build times, this may be an issue.

1. Is this a valid concern? If not, why not?
2. Is this robust when hacking offline?

**Comment by kevgo** [comment]
In my thinking, Git Town values developer productivity (through less conflicts) over trying to save CI capacity. It assumes sufficient CI capacity available on a project (parallelized if needed, so that different developers don't get in each others way), and it aggressively keeps branches in sync with the rest of the world (e.g. "git sync --all" often triggers dozens of CI builds). Computing is cheap these days, developers are expensive, so when choosing to optimize for something, lets optimize for the resource that costs several thousand dollars per month, and not the one that is either free (for open source), or costs maybe a hundred to two hundred per month. 

I'd also make the case that even a new branch should ideally be built before there are any changes made to it. That's especially true if the master isn't always green (which we still can't guarantee). If I create a new branch, and start coding in it, I might want to be notified as soon as possible that something in a long-running test suite doesn't work on my branch. Not everybody checks the status of `master` before cutting a new branch. I certainly don't. 

Also, recognizing that a fresh branch points to an already tested commit on master, and as a result skipping to test it again, is a responsibility of the CI server, not the developer who cuts a ton of branches every day. The CI server has all the information to make that decision. So let's not make this decision the responsibility of Git Town.

**Comment by kevgo** [comment]
When hacking offline, all remote operations are off, so this one as well. I think that's robust and consistent, since the next `git sync` will sync the local repo with the remote one.

**Comment by allewun** [comment]
> In my thinking, Git Town values developer productivity (through less conflicts) over trying to save CI capacity. It assumes sufficient CI capacity available on a project (parallelized if needed, so that different developers don't get in each others way), and it aggressively keeps branches in sync with the rest of the world (e.g. "git sync --all" often triggers dozens of CI builds). Computing is cheap these days, developers are expensive, so when choosing to optimize for something, lets optimize for the resource that costs several thousand dollars per month, and not the one that is either free (for open source), or costs maybe a hundred to two hundred per month. 

CircleCI doesn't support parallel builds for iOS (yet), so from time to time it becomes the bottleneck for us. I recognize this as an edge case, but for my particular workflow having to cancel Circle builds from time to time to unblock the build queue makes a previously cheap command a bit more expensive. Just one use-case I wanted to share.

> Also, recognizing that a fresh branch points to an already tested commit on master, and as a result skipping to test it again, is a responsibility of the CI server, not the developer who cuts a ton of branches every day. The CI server has all the information to make that decision. So let's not make this decision the responsibility of Git Town.

I agree, though I don't believe CircleCI has this ability.


@kevgo Overall I see the value and the benefits this provides. However, maybe I'm more conservative about coupling these commands. By coupling them, it seems you make the 90% use-case easier, but make the 10% use-case more difficult. With with the commands uncoupled (`git-hack` and `git push`), you allow for more flexible use-cases at the expense of a tiny bit more to type. Not against this feature, but just wanted to throw my thoughts out there.

**Comment by charlierudolph** [comment]
How about we make this a configuration option or whether or not to push up branches which do not have a unique commit which isn't in the main branch?

**Comment by kevgo** [comment]
If we make the push portion of `git hack` optional, what about `git sync` and `git ship`? I could come up with all sorts of edge case scenarios for those as well. Maybe I want to sync and then ship right away, so let's omit the push in the sync. Maybe I want to ship 2 branches in a row, so let's omit the push of the first ship (but not the second). 

In the end, supporting this type of optimizations adds a ton of complexity, configuration effort, and learning curve to the product, for no good benefit. If the CI server is too slow, or not intelligent enough, the solution should be to improve the CI server, make the test suite faster, or run only smoke tests for normal pushes and the full suite only for shipping. Not compromise the development workflow. 

Keep in mind that #633 requires this as well. If we always push branches up, a missing remote branch becomes a powerful signal. In this case it means somebody else has shipped or deleted the branch. This means we should ask the user whether she wants to also delete the branch on her machine, or push it up again. If we don't auto-push branches up to the repo, we don't know about that. Then we either push too many outdated branches up to the repo (as we do now), or ask a lot of stupid questions to our users (once per branch).

**Comment by kevgo** [comment]
Another thing that bothers me almost every day that this PR fixes: A newly created branch behaves differently before and after the first git-sync. Before I cannot run `git push`, afterwards I can. 

This impacts me almost every day. I rely on CI to run the test suite for me and find issue is branches that I'm working on. I therefore `git push` frequently. With the current `git hack` setup that doesn't push to the remote, I always have to run a `git sync` right after the `git hack`. 

This issue removes that inconsistency.

**Comment by kevgo** [comment]
Here is another use case: When branching off another branch (instead of main), possibly from another developer, I want CI to tell me right away if my new branch builds correctly or not. Not when I'm knee-deep coding and starting to sync my changes to the repo.

**Comment by charlierudolph** [comment]
As we discussed in the office, #663 doesn't actually require as we can differentiate between branches with no remote and those with a remote that got deleted. 

@allewun You said this is better in 90% of the cases but bad in 10%. Thus how do you feel about a configuration option that allows you to turn this off?

**Comment by allewun** [comment]
Do you mean something like `git hack --local`? If that doesn't add too much complexity, I would be ok with it.

**Comment by charlierudolph** [comment]
I'm thinking more of a configuration option that can be set with the `git town hack-push <value>`. We can default it to yes and thus you can easily turn it off

**Comment by kevgo** [comment]
Given that there is another option to determine missing remote branches, the only motivator for this (at least for me) is the improved consistency of new branches pre and post sync. I.e. the ability to run `git push` at any time, and the ability to see unpushed changes at any time now.

Given that, I'm fine with a config option as @charlierudolph described being added, if there is a demand for it.

**Comment by allewun** [comment]
@kevgo @charlierudolph are we still good with a configuration option? I've actually been using git-town revision c5e10c4 this whole time to satisfy my use-case... it'd be nice to use the latest and greatest. :grin:

**Comment by charlierudolph** [comment]
I'm still good with a config option. Want to create a new issue?

**Comment by kevgo** [comment]
Yes, sounds good to me.
